networkit/cpp/embedding/LearnEmbedding.cpp

```diff
@@ -0,0 +1,221 @@
+/*
+ * LearnEmbeddings.cpp
+ *
+ *  Created on: 06.07.2020
+ *      Author: Klaus Ahrens  <ahrens@informatik.hu-berlin.de>
+ *
+ *  adapted and reimplemented from node2vec
+ *  part of snap [https://github.com/snap-stanford/snap]
+ *  Copyright (c) 2007-2019, Jure Leskovec (under BSD license)
+ *
+ *  see [https://arxiv.org/pdf/1607.00653v1.pdf]
+ *
+ */
+
+// networkit-format
+
+#include <algorithm>
+#include <fstream>
+#include <unordered_map>
+#include <vector>
+
+#include <networkit/auxiliary/Random.hpp>
+
+#include "AliasSampler.hpp"
+#include "BiasedRandomWalk.hpp"
+#include "LearnEmbedding.hpp"
+
+// Code from https://github.com/nicholas-leonard/word2vec/blob/master/word2vec.c
+// Customized for SNAP and node2vec
+
+namespace NetworKit {
+namespace Embedding {
+
+// Number of negative samples. Value taken from original word2vec code.
+constexpr int negSamN = 5;
+
+constexpr int maxExp = 6;
+
+// Learning parameters for SGD. Values taken from original word2vec code.
+constexpr double startAlpha = 0.025;
+constexpr int refreshAlphaCount = 10000;
+constexpr double minAlpha = startAlpha * 0.0001;
+constexpr double Pwr = 0.75;
+
+using Vocab = std::vector<count>;
+
+Vocab learnVocab(AllWalks &allWalks, count nn) {
+    Vocab vocab(nn);
+    for (auto &walk : allWalks) { // for every walk starting from this node
+        for (auto &node : walk) { // for every node in this walk
+            ++vocab[node];
+        }
+    }
+    return vocab;
+}
+
+// Precompute unigram table using alias sampling method
+AliasSampler vocabSampler(Vocab &vocab) {
+    double trainWordsPow = 0;
+    auto sz = vocab.size();
+    std::vector<float> probV(sz); // NOT ... {sz}
+    for (count i = 0; i < sz; ++i) {
+        double e = std::pow(vocab[i], Pwr);
+        probV[i] = e;
+        trainWordsPow += e;
+    }
+    std::transform(probV.begin(), probV.end(), probV.begin(),
+                   [&trainWordsPow](float f) { return f / trainWordsPow; });
+
+    AliasSampler as(sz);
+    as.unigram(probV);
+    return as;
+}
+
+// Initialize negative embeddings
+Embeddings initNegEmb(Vocab &vocab, int dimensions) {
+    return Embeddings{vocab.size(), std::vector<float>(dimensions)};
+}
+
+// Initialize positive embeddings
+Embeddings initPosEmb(Vocab &vocab, int dimensions) {
+    Embeddings emb{vocab.size(), std::vector<float>(dimensions)};
+
+    for (index i = 0; i < emb.size(); ++i) {
+        std::generate(emb[i].begin(), emb[i].end(),
+                      [&]() { return (uniform_real() - 0.5) / dimensions; });
+    }
```

Comment:
It looks like that each iteration of this loop modifies only the i-th element. Can this loop be parallelized? Of course, we need to make sure that `uniform_real()` is thread-safe, one idea to do this is to store a random generator per thread.